### PR TITLE
[Plugin API] Feat: expose guest's favourite ride (ID) to plugin API

### PR DESCRIPTION
**Pull request by mrmagic2020**
Add a new writeable property `favouriteRide: number` to `interface Guest`. The changes are tested locally.

### CODE REVIEW

File: src/openrct2/scripting/bindings/entity/ScGuest.cpp
Comment (spacek531) [inline src/openrct2/scripting/bindings/entity/ScGuest.cpp:514]: The nullptr case should return RideId::GetNull().toUnderlying()

**Review by spacek531 — CHANGES_REQUESTED** [review]
Only one comment to make, when that is changed this is good 👍

**Review by spacek531 — APPROVED** [review]

File: distribution/openrct2.d.ts
Comment (Basssiiie) [inline distribution/openrct2.d.ts:2881]: Can you make it actually return `null` when a guest has no favourite ride? You can look at some other properties that can return both a number and null.
```suggestion
        favouriteRide: number | null;
```

File: src/openrct2/scripting/bindings/entity/ScGuest.cpp
Comment (Basssiiie) [inline src/openrct2/scripting/bindings/entity/ScGuest.cpp:523]: Similarly, IMHO the favourite ride should reset when `null` is assigned.

**Review by Basssiiie — CHANGES_REQUESTED** [review]

File: src/openrct2/scripting/bindings/entity/ScGuest.cpp
Comment (mrmagic2020) [reply src/openrct2/scripting/bindings/entity/ScGuest.cpp:523]: Should I use `DukValue` for function return & param type then? That seems to be what similar properties do.

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/entity/ScGuest.cpp:523]: Yes. 🙂

**Comment by mrmagic2020** [comment]
image

File: src/openrct2/scripting/ScriptEngine.h
Comment (spacek531) [inline src/openrct2/scripting/ScriptEngine.h:50]: Develop branch currently has API 89, this needs to be rebased and updated to 90.

**Comment by mrmagic2020** [comment]
@spacek531 Is a check like `value.asuint() < GetGameState().Rides.size()` necessary for `void ScGuest::favouriteRide_set(const DukValue& value)`?

**Comment by spacek531** [comment]
Yes, it needs to be capped to the length of the rides list. I would compare it to `OpenRCT2::Limits::MaxRidesInPark` instead and set to null if higher.

Because elements are not shifted in the vector when a ride is deleted, there can be "holes" in the vector. When deleting a ride, any guests with that ride as their favorite have their favorite set to null. Setting the guest's favourite ride to the `id` of the ride at the index value in `GameState_t.Rides` will be a one-liner no-branches way of automatically filtering out null rides.

**Comment by mrmagic2020** [comment]
> Yes, it needs to be capped to the length of the rides list. I would compare it to `OpenRCT2::Limits::MaxRidesInPark` instead and set to null if higher.
> 
> Because elements are not shifted in the vector when a ride is deleted, there can be "holes" in the vector. When deleting a ride, any guests with that ride as their favorite have their favorite set to null. Setting the guest's favourite ride to the `id` of the ride at the index value in `GameState_t.Rides` will be a one-liner no-branches way of automatically filtering out null rides.

I had a look into `Ride.cpp`, especially the `GetRide()` function. I based the validation logic on that.
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/ride/Ride.cpp#L219~L242

image

**Review by spacek531 — APPROVED** [review]

**Review by Gymnasiast — COMMENTED** [review]
Code looks fine, but I’ll hold off merging until @Basssiiie can give his approval too.

**Review by Basssiiie — APPROVED** [review]
Tested it and it all LGTM! Nice work! 🙂

<details>
<summary>Click here for the script I used for testing it.</summary>

```js
registerPlugin({
	name: "Test guest favourite ride",
	version: "1",
	authors: ['Basssiiie'],
	targetApiVersion: 82,
	type: "local",
	licence: "MIT",
	main()
	{
		ui.registerMenuItem("Test guest favourite ride", function()
		{
			ui.activateTool({
				id: "test-guest-favourite-ride",
				cursor: "cross_hair",
				onDown: function(args)
				{
					var id = args.entityId;
					if (id == undefined) return;

					var guest = map.getEntity(id);
					if (guest.type !== "guest") return;

					var rides = map.rides.slice().sort(function(l,r){return l.name.localeCompare(r.name)});
					var window = ui.openWindow({
						classification: "test-guest-favourite-ride",
						title: "Guest favourite ride",
						width: 210,
						height: 100,
						onUpdate: function()
						{
							var label = window.findWidget("guest-ride");
							var rideId = guest.favouriteRide;
							var ride = rides.filter(function(r){return r.id == rideId})[0];
							if (ride)
							{
								label.text = "Favourite ride: " + ride.name + " (" + guest.favouriteRide + ")";
							}
							else
							{
								label.text = "Favourite ride: null";
							}
						},
						widgets: [
							{
								type: "label",
								x: 5,
								y: 20,
								width: 200,
								height: 15,
								text: "Name: " + guest.name
							},
							{
								name: "guest-ride",
								type: "label",
								x: 5,
								y: 40,
								width: 200,
								height: 15,
							},
							{
								type: "dropdown",
								x: 5,
								y: 60,
								width: 200,
								height: 15,
								items: rides.map(function(r){return r.name}),
								onChange: function(index)
								{
									guest.favouriteRide = rides[index].id;
								}
							},
							{
								type: "button",
								text: "Reset favourite ride",
								x: 5,
								y: 80,
								width: 200,
								height: 15,
								onClick: function()
								{
									guest.favouriteRide = null;
								}
							},
						]
					})
				}
			})
		})
	}
});
```
</details>

File: distribution/changelog.txt
Comment (Gymnasiast) [inline distribution/changelog.txt:9]: No typewriter quotes:

```suggestion
- Feature: [#22087] [Plugin] Expose guests’ favourite rides to the plugin API.
```

**Pull request merged**
